### PR TITLE
[WIP] Adding support for subcollections and related subresources

### PR DESCRIPTION
**Pull request by abellotti**
- Dynamically driven by the "subcollections" exposed via OPTIONS /api/:collection
- supports queries and actions

  Queries:
  miq.vms.find(166).tags.collect(&:name)
  miq.vms.find(166).tags.select(:categorization).collect(&:categorization)

  Subcollection Actions:
  miq.vms.find(166).tags.assign(:name => "/managed/location/ny")
  miq.vms.find(166).tags.assign([{:name => "/managed/location/chicago"}, {:name => "/managed/cc/001"}])

  Subcollection resource actions:
  miq.vms.find(166).tags.find(32).unassign
  miq.vms.find(166).tags.where(:name => "/managed/location/*").collect(&:unassign)

Fixes: #31
Fixes: #32

### CODE REVIEW

**Comment by abellotti** [comment]
WIP - Still TODO:

- more testing
- DRY
- specs

**Comment by abellotti** [comment]
@Fryguy subcollection support + lotsa tests added

**Comment by abellotti** [comment]
ping @Fryguy for review, will be needed shortly for Automate work.  Thanks.

**Comment by abellotti** [comment]
Ping @Fryguy we need this for Automate Workspace work by @mkanoor

File: lib/manageiq/api/client/mixins/collection_action_mixin.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
+module CollectionActionMixin
+  include ActionMixin
+
+  ACTIONS_RETURNING_RESOURCES = %w(create query).freeze
+
+  def exec_action(name, *args, &block)
```
Comment (Fryguy) [inline lib/manageiq/api/client/mixins/collection_action_mixin.rb]: Should any of these be private?

File: lib/manageiq/api/client/mixins/collection_action_mixin.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+module CollectionActionMixin
```
Comment (Fryguy) [inline lib/manageiq/api/client/mixins/collection_action_mixin.rb]: This needs to be namespaced

File: lib/manageiq/api/client/mixins/queryable_mixin.rb [hunk 3]
```diff
@@ -0,0 +1,95 @@
+module QueryableMixin
```
Comment (Fryguy) [inline lib/manageiq/api/client/mixins/queryable_mixin.rb]: Also needs namespace

File: lib/manageiq/api/client/mixins/resource_action_mixin.rb [hunk 4]
```diff
@@ -0,0 +1,24 @@
+module ResourceActionMixin
```
Comment (Fryguy) [inline lib/manageiq/api/client/mixins/resource_action_mixin.rb]: Ditto

File: lib/manageiq/api/client/subcollection.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        end
+
+        def self.subclass(name)
+          name = name.camelize
+
+          if const_defined?(name, false)
+            const_get(name, false)
+          else
+            const_set(name, Class.new(self))
+          end
```
Comment (Fryguy) [inline lib/manageiq/api/client/subcollection.rb]: It feels like there is a lot of duplication with collection.

File: lib/manageiq/api/client/mixins/action_mixin.rb [hunk 6]
```diff
@@ -1,4 +1,4 @@
-module ActionMixin
+module ManageIQ::API::ActionMixin
```
[the remark this answers is not in the document]
Comment (abellotti) [reply lib/manageiq/api/client/mixins/action_mixin.rb]: good point 👍  will update all.

File: lib/manageiq/api/client/mixins/queryable_mixin.rb [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    end
+    filters
+  end
+
+  def order_parameters_from_query_relation(option)
+    query_relation_option =
+      if option.kind_of?(Array)
+        option.each_with_object({}) { |name, hash| hash[name] = "asc" }
+      else
+        option.dup
```
Comment (Fryguy) [inline lib/manageiq/api/client/mixins/queryable_mixin.rb]: What is this dup for if you're going to loop over it anyway?

File: lib/manageiq/api/client/mixins/queryable_mixin.rb [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    end.flatten
+  end
+
+  def order_parameters_from_query_relation(option)
+    query_relation_option =
+      if option.kind_of?(Array)
+        option.each_with_object({}) { |name, hash| hash[name] = "asc" }
+      else
+        option
+      end
```
Comment (Fryguy) [inline lib/manageiq/api/client/mixins/queryable_mixin.rb]: I feel like you can remove this completely if you just deal with defaulting nil values below.

```ruby
[1] pry(main)> option = {"a" => "asc", "b" => "desc"}
=> {"a"=>"asc", "b"=>"desc"}
[2] pry(main)> option.each {|k, v| puts "#{k} => #{v || "asc"}"}
a => asc
b => desc
=> {"a"=>"asc", "b"=>"desc"}
[3] pry(main)> option = ["a", "b"]
=> ["a", "b"]
[4] pry(main)> option.each {|k, v| puts "#{k} => #{v || "asc"}"}
a => asc
b => asc
=> ["a", "b"]
```

File: lib/manageiq/api/client/mixins/queryable_mixin.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+    res_sort_by = []
+    res_sort_order = []
+    query_relation_option.each do |sort_attr, sort_order|
+      res_sort_by << sort_attr
+      sort_order =
+        case sort_order
+        when /^asc/i  then "asc"
```
Comment (Fryguy) [inline lib/manageiq/api/client/mixins/queryable_mixin.rb]: so this could be `when /^asc/i, nil then "asc"`

**Comment by Fryguy** [comment]
@abellotti This is a lot to digest `+3,715 −135`...might need to sit with you unless you can break it up into easily digestable PRs

**Comment by abellotti** [comment]
Marked as WIP, I need to revisit Subresource  (/api/providers/1/vms/133  vs just /api/vms/133)

**Comment by abellotti** [comment]
Agreed,  getting slowly morphed/simplified as https://github.com/abellotti/manageiq-api-client/tree/enhanced_subcollection_support.   will create new PR when in good shape.  By simply
using resource.rb (eliminating subresource.rb and possibly subcollection.rb), this would now allow for deeply nested resource/subcollection/action calls as we'd reset at the primary CI URL where actions are provided. So allowing for the following:

```
miq.providers.find_by(:name => "test-provider).vms.find_by(:name => "test-vm").snapshots.collect(&:name)
```

**Comment by abellotti** [comment]
closing in preference of https://github.com/ManageIQ/manageiq-api-client/pull/80

**Pull request closed without merging**